### PR TITLE
Implement Stripe and Subscription modules

### PR DESCRIPTION
**Pull request by doomspork**
This contains the domain logic surrounding subscriptions, database persistence, and interaction with Stripe's API.

### CODE REVIEW

**Comment by doomspork** [comment]
Making use of both `bypass` _and_ `Mox`. I think this could be a blog post and good example project for others

**Comment by btkostner** [comment]
So this will have webhooks for stripe to update our local database. It will take a stripe customer and source and create (or delete) the subscription.

1) If a user cancels the subscription and then subscribes again, it's a new stripe subscription id?
2) if a user creates a subscription with a card, and then tries to delete that card from his account, what happens?

**Comment by doomspork** [comment]
> So this will have webhooks for stripe to update our local database.

Yes. We'll subscribe to the appropriate events on Stripe so the sponsors database is kept up-to-date.

> It will take a stripe customer and source and create (or delete) the subscription.

Creating a new subscription requires the System76 user ID and the Stripe Customer ID we have on file for them. As it is _right now_ the source isn't necessary because it is attached to the default payment method on the user. If we're providing the UI for selecting different sources on their account, I'll make the necessary API changes.

1. Yeah. Each subscription with Stripe gets a newly generated ID.
2. It will use the default source. If they delete _all_ of their cards. The subscription renewal will fail. That's an event we can listen for. I've been discussing these situations with Maria, we'll be adding email notifications for most if not all of them.

**Comment by doomspork** [comment]
[Sponsors service](https://trello.com/c/3FAufyPE/199-sponsors-service)

**Comment by btkostner** [comment]
And lastly, how do we plan to do auth on this?

**Comment by doomspork** [comment]
[Implement schemas and domain logic for sponsorships](https://trello.com/c/vtLkDCRl/210-implement-schemas-and-domain-logic-for-sponsorships)

File: priv/repo/migrations/20200421143406_create_subscriptions_table.exs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      add :stripe_subscription_id, :string, null: false
+
+      timestamps()
+    end
+
+    create unique_index(:subscriptions, [:stripe_subscription_id])
+
+    create table(:invoices) do
+      add :subscription_id, references(:subscriptions)
+      add :stripe_invoice_id, :string, null: false
```
Comment (btkostner) [inline priv/repo/migrations/20200421143406_create_subscriptions_table.exs:17]: unique index on this?

**Review by btkostner — APPROVED** [review]

File: priv/repo/migrations/20200421143406_create_subscriptions_table.exs (hunk 1 above)
Comment (doomspork) [reply priv/repo/migrations/20200421143406_create_subscriptions_table.exs:17]: Good idea. Done.

**Pull request merged**